autopush/router/apns2.py

```diff
@@ -129,33 +133,40 @@ def send(self, router_token, payload, apns_id,
             headers['apns-expiration'] = str(exp)
         url = '/3/device/' + router_token
         connection = self._get_connection()
-        try:
-            # request auto-opens closed connections, so if a connection
-            # has timed out or failed for other reasons, it's automatically
-            # re-established.
-            stream_id = connection.request(
-                'POST', url=url, body=body, headers=headers)
-            # get_response() may return an AttributeError. Not really sure
-            # how it happens, but the connected socket may get set to None.
-            # We'll treat that as a premature socket closure.
-            response = connection.get_response(stream_id)
-            if response.status != 200:
-                reason = json.loads(response.read().decode('utf-8'))['reason']
-                raise RouterException(
-                    "APNS Transmit Error {}:{}".format(response.status,
-                                                       reason),
-                    status_code=502,
-                    response_body="APNS could not process "
-                                  "your message {}".format(reason),
-                    log_exception=False
-                )
-        except (HTTP20Error, AttributeError):
-            connection.close()
-            raise
-        finally:
-            # Returning a closed connection to the pool is ok.
-            # hyper will reconnect on .request()
-            self._return_connection(connection)
+        attempt = 0
+        while True:
+            try:
+                # request auto-opens closed connections, so if a connection
+                # has timed out or failed for other reasons, it's automatically
+                # re-established.
+                stream_id = connection.request(
+                    'POST', url=url, body=body, headers=headers)
+                # get_response() may return an AttributeError. Not really sure
+                # how it happens, but the connected socket may get set to None.
+                # We'll treat that as a premature socket closure.
+                response = connection.get_response(stream_id)
+                if response.status != 200:
+                    reason = json.loads(
+                            response.read().decode('utf-8'))['reason']
+                    raise RouterException(
+                        "APNS Transmit Error {}:{}".format(response.status,
+                                                           reason),
+                        status_code=502,
+                        response_body="APNS could not process "
+                                      "your message {}".format(reason),
+                        log_exception=False
+                    )
+                break
+            except HTTP20Error:
+                connection.close()
+                attempt += 1
+                if attempt < self._max_retry:
+                    continue
```

Comment:
The _return_connection in the finally block's a problem now because we only call get_connection once (outside of the loop).

So in the case here of the loop continue we're potentially returning it multiple times (per each iteration).

Probably wanna set a flag here to disable returning in the finally block, or could even pull the loop body into a separate method

Reply:
Crap, you're right. good catch!

Reply:
And the proper fix is to move _get_connection inside of the loop since we need to pull a new one if it closes or fails.